### PR TITLE
[IMP] base: prefetch actions models in batch during /web loading

### PR DESCRIPTION
**Pull request by Feyensv**
Gain of 200+ queries in a db with test_main_flows & dependencies 
installed, noticeable reduction of time taken by `load_menus` part.

Note that this only impacts first request since this method is cached.  
Unless user groups (or debug status) are different, the cached results 
are returned.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_ui_menu.py [hunk 1]
```diff
@@ -93,15 +93,31 @@ def _visible_menu_ids(self, debug=False):
...
+            for model in MODEL_FIELD
+        }
+        for menu in action_menus:
+            action = menu.action
+            if action._name in MODEL_FIELD:
+                actions_collector[action._name] += action
+        for model, fname in MODEL_FIELD.items():
+            actions = actions_collector[model]
+            if actions:
+                actions.sudo().mapped(fname)
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_menu.py:114]: Another approach is to determine the "prefetch set" of actions by category
```suggestion
        # determine the ids to prefetch by category
        prefetch_ids = defaultdict(list)
        for action in action_menus.mapped('action'):
            prefetch_ids[action._name].append(action.id)
```
and then in the loop below, add the line
```python
action = action.with_prefetch(prefetch_ids[action._name])
```

File: odoo/addons/base/models/ir_ui_menu.py [hunk 2]
```diff
@@ -93,15 +93,31 @@ def _visible_menu_ids(self, debug=False):
 
         # process action menus, check whether their action is allowed
         access = self.env['ir.model.access']
-        MODEL_GETTER = {
-            'ir.actions.act_window': lambda action: action.res_model,
-            'ir.actions.report': lambda action: action.model,
-            'ir.actions.server': lambda action: action.model_id.model,
+        MODEL_FIELD = {
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_menu.py:96]: That name is not clear to me.  If we call `action._name` its "category" or "type", then I suggest `MODEL_BY_CATEGORY` or `MODEL_BY_TYPE`.

File: odoo/addons/base/models/ir_ui_menu.py [hunk 3]
```diff
@@ -93,15 +93,31 @@ def _visible_menu_ids(self, debug=False):
...
+            if actions:
+                actions.sudo().mapped(fname)
+
         for menu in action_menus:
-            get_model = MODEL_GETTER.get(menu.action._name)
-            if not get_model or not get_model(menu.action) or \
-                    access.check(get_model(menu.action), 'read', False):
+            action = menu.action
+            action_model = action._name
+            model = action_model in MODEL_FIELD and action[MODEL_FIELD[action_model]]
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_menu.py:119]: `model` → `model_name`

**Review by rco-odoo — APPROVED** [review]
@robodoo r+

**Comment by yelizariev** [comment]
We can gain another 300+ queries. Please check: https://github.com/odoo/odoo/pull/90710

**Comment by yelizariev** [comment]
We should do something with reviewing. I suggested a similar trick 6 months before this PR https://github.com/odoo/odoo/commit/c923072e16f0fcc0c4f6c9da22c9aa3c235d2f9a
https://www.odoo.com/web#id=2614441&cids=1&menu_id=4720&action=4043&model=project.task&view_type=form

**Pull request closed without merging**